elements/pl-python-variable/pl-python-variable.py

```diff
@@ -1,74 +1,122 @@
         element, "show-dimensions", SHOW_DIMENSIONS_DEFAULT
     )
+    varname = pl.get_string_attrib(element, "params-name")
+    indent = pl.get_integer_attrib(element, "indent", INDENT_DEFAULT)
+    depth = pl.get_integer_attrib(element, "depth", DEPTH_DEFAULT)
+    width = pl.get_integer_attrib(element, "width", WIDTH_DEFAULT)
+    compact = pl.get_boolean_attrib(element, "compact", COMPACT_DEFAULT)
+    sort_dicts = pl.get_boolean_attrib(element, "sort-dicts", SORT_DICTS_DEFAULT)
+    # TODO this is a python 3.10-only addition, hold off on adding this till non-US servers are upgraded.
+    # underscore_numbers = pl.get_boolean_attrib(
+    #    element, "underscore-numbers", UNDERSCORE_NUMBERS_DEFAULT
+    # )
 
     if varname not in data["params"]:
-        raise Exception("Could not find {} in params!".format(varname))
-
-    var_out = data["params"][varname]
-    html = ""
-    var_type = "text"
-
-    # determine the type of variable to render
-    if isinstance(var_out, dict) and "_type" in var_out:
-        if not force_text:
-            var_type = var_out["_type"]
-        var_out = pl.from_json(var_out)
-
-    # render the output variable
-    if var_type == "dataframe":
-        html += var_out.to_html(
-            header=show_header, index=show_index, classes=["pl-python-variable-table"]
+        raise KeyError(
+            f'Could not find {varname} in params. Please make sure to set params-name="{varname}" in the element.'
         )
-        if show_dimensions:
-            html += '<p class="pl-python-variable-table-dimensions">{} rows x {} columns</p>'.format(
-                str(var_out.shape[0]), str(var_out.shape[1])
+
+    json_object = data["params"][varname]
+
+    if isinstance(json_object, dict) and "_value" in json_object:
+        # Note this will always work, since if from_json can't convert the object, it does nothing.
+        var_out = pl.from_json(json_object)
+
+        # Passthrough legacy support for pl-dataframe
+        if isinstance(var_out, pandas.DataFrame) and not force_text:
+            return (
+                f'<pl-dataframe params-name="{varname}" show-header="{show_header}" show-index="{show_index}" '
+                f'show-dimensions="{show_dimensions}" show-python="False"></pl-dataframe>'
             )
+        else:
+            var_string = repr(var_out)
+
     else:
-        no_highlight = pl.get_boolean_attrib(
-            element, "no-highlight", NO_HIGHLIGHT_DEFAULT
-        )
-        prefix = pl.get_string_attrib(element, "prefix", PREFIX_DEFAULT)
-        suffix = pl.get_string_attrib(element, "suffix", SUFFIX_DEFAULT)
+        if isinstance(json_object, str):
+            var_string = repr(json_object)
 
-        text = prefix + repr(var_out) + suffix
-        html += '<pl-code language="python" no-highlight="{}">{}</pl-code>'.format(
-            no_highlight, text
-        )
+        else:
+            var_string = pprint.pformat(
```

Comment:
Maybe change pretty print to only work with lists and dicts and then have everything else use repr? I think those are the only cases we care about being able to print complex nested things anyway.

Reply:
I think Canada is on 3.8, but the only thing missing is the `underscore_numbers` parameter. That's supposed to be in the library in 3.10, but weirdly had issues getting that working. I guess we can try including it later.

Reply:
@coatless the one issue is that tuples and sets can't be json serialized, so they'll never be output types here. It's easy enough to add them in pl.to_json though. I'll still add the types to the instance check.

Reply:
@eliotwrobson effort is worth it to serialize these structures. 

Not a fun experience when trying to write questions about Python structures. The element gets a bit invalidated if you need to write a custom string pass to `pl-code` with printing. 

Let me think about serialization. I think this PR is on the bottom of your todo list? I probably can have a solution on Tuesday of next week.

Reply:
Gotcha. No worries on waiting, I can knock off another one in the interim.

Reply:
Since this has been approved with the current semantics and features, I think it's fine to add the ability to print sets / tuples in later.

Reply:
Added as an issue #7000